tests/returntypes/test_tensorflow_qnode_shot_vector.py

```diff
@@ -0,0 +1,581 @@
+        num_copies = sum(
+            [1 for x in shots if isinstance(x, int)] + [x[1] for x in shots if isinstance(x, tuple)]
+        )
+        assert jac.shape == (num_copies, 5, 2)
+
+    def test_hessian_expval_multiple_params(self, dev_name, diff_method, gradient_kwargs, shots):
+        """The hessian of a single measurement with multiple params return a tuple of arrays."""
+        dev = qml.device(dev_name, wires=2, shots=shots)
+
+        par_0 = tf.Variable(0.1, dtype=tf.float64)
+        par_1 = tf.Variable(0.2, dtype=tf.float64)
+
+        @qnode(dev, diff_method=diff_method, interface="tf", max_diff=2, **gradient_kwargs)
+        def circuit(x, y):
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        with tf.GradientTape() as tape1:
+            with tf.GradientTape(persistent=True) as tape2:
+                res = circuit(par_0, par_1)
+                res = qml.math.stack(res)
+
+            jac = tape2.jacobian(res, (par_0, par_1), experimental_use_pfor=False)
+            jac = qml.math.stack(jac)
+
+        hess = tape1.jacobian(jac, (par_0, par_1))
+
+        assert isinstance(hess, tuple)
+        num_copies = sum(
+            [1 for x in shots if isinstance(x, int)] + [x[1] for x in shots if isinstance(x, tuple)]
+        )
+        assert len(hess) == 2
+        for h in hess:
+            assert isinstance(h, tf.Tensor)
+            assert h.shape == (2, num_copies)
+
+    def test_hessian_expval_multiple_param_array(
+        self, dev_name, diff_method, gradient_kwargs, shots
+    ):
+        """The hessian of single measurement with a multiple params array return a single array."""
+        dev = qml.device(dev_name, wires=2, shots=shots)
+
+        params = tf.Variable([0.1, 0.2], dtype=tf.float64)
+
+        @qnode(dev, diff_method=diff_method, interface="tf", max_diff=2, **gradient_kwargs)
+        def circuit(x):
+            qml.RX(x[0], wires=[0])
+            qml.RY(x[1], wires=[1])
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        with tf.GradientTape() as tape1:
+            with tf.GradientTape(persistent=True) as tape2:
+                res = circuit(params)
+                res = qml.math.stack(res)
+
+            jac = tape2.jacobian(res, params, experimental_use_pfor=False)
+
+        hess = tape1.jacobian(jac, params)
+
+        assert isinstance(hess, tf.Tensor)
+        num_copies = sum(
+            [1 for x in shots if isinstance(x, int)] + [x[1] for x in shots if isinstance(x, tuple)]
+        )
+        assert hess.shape == (num_copies, 2, 2)
+
+    def test_hessian_probs_expval_multiple_params(
+        self, dev_name, diff_method, gradient_kwargs, shots
+    ):
+        """The hessian of multiple measurements with multiple params return a tuple of arrays."""
+
+        # override shots to speed up the test
```

Comment:
What was time consuming? Having two different shot vectors or having many shots in total?

The thing with the below `shots` definition is that it doesn't contain a tuple - an edge case which would be great to test for.

Reply:
The Hessians are in general very time-consuming, but even more so if there are a lot of copies in the shot vector. I can create a new `pytest.mark.parametrize` marker just for the Hessians and include a tuple in the shot vector.

Reply:
Using the `slow` marker can also be useful here.

Reply:
Added `slow` markers for the Hessian tests.